### ISSUE TITLE
bug with dynamic skip functionality

### ISSUE THREAD

**Issue by ghughal** [mochajs/mocha#1604]
I'm getting error with using dynamic skip() functionality with asyn function. Is this a bug or I'm doing something wrong?

``` javascript
    it('should skip after callback', function(done){
        var self = this;
        var asyncFunc = function(cb) {
            setTimeout(cb, 1000);
        };

        asyncFunc(function() {
            self.skip();
            throw new Error('never thrown');
        });
    })
```

Here's error I'm getting:

```
  1) skip in test should skip after callback:
     Error: the object {
  "message": [undefined]
  "uncaught": true
} was thrown, throw an Error :)
      at Runner.fail (/usr/local/lib/node_modules/mocha/lib/runner.js:207:11)
      at Runner.uncaught (/usr/local/lib/node_modules/mocha/lib/runner.js:580:8)
      at process.uncaught (/usr/local/lib/node_modules/mocha/lib/runner.js:609:10)
      at process.EventEmitter.emit (events.js:95:17)
      at process._fatalException (node.js:272:26)
```

**Comment by a8m** [comment]
It's seems like a bug.
I started to play with it, and I'll update later.
Thx @ghughal

**Comment by ghughal** [comment]
Thanks @a8m

**Comment by danielstjules** [comment]
@ghughal @a8m I don't think this behavior's ever been supported, but I submitted a possible solution for allowing it: https://github.com/mochajs/mocha/pull/1618

**Comment by ghughal** [comment]
Thank you @danielstjules

**Comment by ScottFreeCode** [comment]
This should be fixed by #2335; can anyone confirm?

**Comment by maggiesavovska** [comment]
I'm getting the following error: 
"Converting circular structure to JSON"

`it('should skip async test', function(done){
            var self = this;
            return driver.url('https://www.google.com').then(function(result){
                self.skip();

            }).catch(function(err){console.log('error', err.message)});

        });`

**Comment by ScottFreeCode** [comment]
I just checked with the various versions of Mocha and confirmed that this issue from 2.x was fixed in 3.0. I also checked the PR and it added regression tests for this specific case, so this shouldn't come up again either. 🎆 

@maggiesavovska I was unable to get a circular JSON error out of `skip()` in 2.x, 3.0 or current, using that particular test with `Promise.resolve` substituted for `driver.url` (or with any variations I could think of). Can you confirm whether you can still get that result in current Mocha and if so open a new issue including which library you're using for `driver`? (It also might be a duplicate of #2598, but we may as well take a look at your case before we reach such a conclusion.)

**Comment by maggiesavovska** [comment]
@ScottFreeCode: Yes, it's fixed in mocha@3.1.0. Thank you!
